### PR TITLE
Add Localized Items to GraphQL

### LINKED ISSUE #5268: GaphQL localized contentItem
Is there a way to get the ID of the same item in different language if it's available and not only the current  localization proprieties 
Example; i need english version of this event: 
image

### PR DESCRIPTION
**Pull request by pbros**
Add LocalizationSet Filter on Localization Part
Add Localized Content Items with culture filter on Localization Part

Fixes #5268

### CODE REVIEW

File: src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs [hunk 1]
```diff
@@ -13,6 +18,27 @@ public LocalizationQueryObjectType(IStringLocalizer<LocalizationQueryObjectType>
...
             Field(x => x.LocalizationSet).Description(S["The localization set for your content item."]);
+
+            Field<ListGraphType<ContentItemType>, IEnumerable<ContentItem>>()
+                .Name("Localizations")
+                .Description(S["The localizations of the content item."])
+                .Argument<StringGraphType, string>("culture", "the culture of the content item", string.Empty)
+                .Resolve(x =>
+                {
+                    var culture = x.GetArgument<string>("culture");
+
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs]: Remove the extra line

File: src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs [hunk 2]
```diff
@@ -13,6 +18,27 @@ public LocalizationQueryObjectType(IStringLocalizer<LocalizationQueryObjectType>
...
+                    var contentLocalizationManager = context.ServiceProvider.GetService<IContentLocalizationManager>();
+
+                    if (!string.IsNullOrWhiteSpace(culture))
+                    {
+                        return new List<ContentItem>
+                        {
+                            contentLocalizationManager.GetContentItemAsync(x.Source.LocalizationSet, culture).GetAwaiter().GetResult()
+                        };
+                    }
+                    return contentLocalizationManager.GetItemsForSetAsync(x.Source.LocalizationSet).GetAwaiter().GetResult();
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs]: Add new line before

File: src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs [hunk 3]
```diff
@@ -13,6 +18,27 @@ public LocalizationQueryObjectType(IStringLocalizer<LocalizationQueryObjectType>
...
+                    var culture = x.GetArgument<string>("culture");
+
+                    var context = (GraphQLContext)x.UserContext;
+                    var contentLocalizationManager = context.ServiceProvider.GetService<IContentLocalizationManager>();
+
+                    if (!string.IsNullOrWhiteSpace(culture))
+                    {
+                        return new List<ContentItem>
+                        {
+                            contentLocalizationManager.GetContentItemAsync(x.Source.LocalizationSet, culture).GetAwaiter().GetResult()
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs]: I'm not sure why you didn't use `await here?!!

File: src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs [hunk 4]
```diff
@@ -13,6 +18,27 @@ public LocalizationQueryObjectType(IStringLocalizer<LocalizationQueryObjectType>
...
+                .Description(S["The localizations of the content item."])
+                .Argument<StringGraphType, string>("culture", "the culture of the content item", string.Empty)
+                .Resolve(x =>
+                {
+                    var culture = x.GetArgument<string>("culture");
+
+                    var context = (GraphQLContext)x.UserContext;
+                    var contentLocalizationManager = context.ServiceProvider.GetService<IContentLocalizationManager>();
+
+                    if (!string.IsNullOrWhiteSpace(culture))
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs]: Is this mean that the culture could not be `Invariant`?!!

/cc @Skrypt

File: src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs (hunk 1 above)
Comment (pbros) [reply src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs]: fixed

File: src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs (hunk 2 above)
Comment (pbros) [reply src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs]: fixed

File: src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs (hunk 3 above)
Comment (pbros) [reply src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs]: Hadn't realized there was a ResovleAsync method. Using that now and await

File: src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs (hunk 4 above)
Comment (pbros) [reply src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs]: Added support for Invariant by setting the default value to null and doing a nullcheck instead.

**Comment by sebastienros** [comment]
Looks great, does if provide enough features to solve your requirements?

File: src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs [hunk 5]
```diff
@@ -13,6 +18,27 @@ public LocalizationQueryObjectType(IStringLocalizer<LocalizationQueryObjectType>
 
             Field(x => x.Culture).Description(S["The culture for your content item."]);
             Field(x => x.LocalizationSet).Description(S["The localization set for your content item."]);
+
+            Field<ListGraphType<ContentItemType>, IEnumerable<ContentItem>>()
+                .Name("Localizations")
+                .Description(S["The localizations of the content item."])
+                .Argument<StringGraphType, string>("culture", "the culture of the content item", null)
+                .ResolveAsync( async x =>
```
Comment (carlwoodhouse) [inline src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs]: because your resolving a service that likely uses sql further down the stack please use the `ResolveLockedAsync` extension, this will stop thread safety issues with YesSQL sessions when not using dataloaders

File: src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs [hunk 6]
```diff
@@ -13,6 +18,27 @@ public LocalizationQueryObjectType(IStringLocalizer<LocalizationQueryObjectType>
 
             Field(x => x.Culture).Description(S["The culture for your content item."]);
             Field(x => x.LocalizationSet).Description(S["The localization set for your content item."]);
+
+            Field<ListGraphType<ContentItemType>, IEnumerable<ContentItem>>()
+                .Name("Localizations")
+                .Description(S["The localizations of the content item."])
+                .Argument<StringGraphType, string>("culture", "the culture of the content item", null)
```
Comment (carlwoodhouse) [inline src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs]: it should default to null by default without having to need to specify explicitly (i think heh)

File: src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs [hunk 7]
```diff
@@ -13,6 +18,27 @@ public LocalizationQueryObjectType(IStringLocalizer<LocalizationQueryObjectType>
...
+                .ResolveAsync( async x =>
+                {
+                    var culture = x.GetArgument<string>("culture");
+                    var context = (GraphQLContext)x.UserContext;
+                    var contentLocalizationManager = context.ServiceProvider.GetService<IContentLocalizationManager>();
```
Comment (carlwoodhouse) [inline src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs]: you can shorten this to `x.ResolveServiceProvider().GetService ..`

File: src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs [hunk 8]
```diff
@@ -13,6 +18,27 @@ public LocalizationQueryObjectType(IStringLocalizer<LocalizationQueryObjectType>
...
+
+                    if (culture != null)
+                    {
+                        return new List<ContentItem>
+                        {
+                            await contentLocalizationManager.GetContentItemAsync(x.Source.LocalizationSet, culture)
```
Comment (carlwoodhouse) [inline src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs]: can this return null ? if so we're going to end up with a list with a null contentitem in it, as the field is not marked as nullable if this is the case we need to return an empty enumeration of contentitem

File: src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs [hunk 9]
```diff
@@ -13,6 +18,27 @@ public LocalizationQueryObjectType(IStringLocalizer<LocalizationQueryObjectType>
 
             Field(x => x.Culture).Description(S["The culture for your content item."]);
             Field(x => x.LocalizationSet).Description(S["The localization set for your content item."]);
+
+            Field<ListGraphType<ContentItemType>, IEnumerable<ContentItem>>()
```
Comment (carlwoodhouse) [inline src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs]: this should be probably be `ContentItemInterface`, not `ContentItemType`

**Comment by pbros** [comment]
It does.

@sallenkov what about you?

File: src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationInputObjectType.cs [hunk 10]
```diff
@@ -13,6 +13,7 @@ public LocalizationInputObjectType(IStringLocalizer<LocalizationInputObjectType>
             Description = S["the localization part of the content item"];
 
             AddScalarFilterFields<StringGraphType>("culture", S["the culture of the content item to filter"]);
+            Field(x => x.LocalizationSet).Description(S["the localization set for your content item."]);
```
Comment (carlwoodhouse) [inline src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationInputObjectType.cs]: for consistency this should probably be `the localization set of the content item to filter`

File: src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs [hunk 11]
```diff
@@ -13,6 +18,27 @@ public LocalizationQueryObjectType(IStringLocalizer<LocalizationQueryObjectType>
...
+
+                    if (culture != null)
+                    {
+                        return new List<ContentItem>
+                        {
+                            await contentLocalizationManager.GetContentItemAsync(x.Source.LocalizationSet, culture)
+                        };
+                    }
+
+                    return await contentLocalizationManager.GetItemsForSetAsync(x.Source.LocalizationSet);
```
Comment (carlwoodhouse) [inline src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs]: same as above, can this return null ?

File: src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs (hunk 5 above)
Comment (pbros) [reply src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs]: fixed

File: src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs (hunk 6 above)
Comment (pbros) [reply src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs]: I had added it to be explicit because of the `if (culture != null)` below, but I removed it

File: src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs (hunk 7 above)
Comment (pbros) [reply src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs]: done

File: src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs (hunk 8 above)
Comment (pbros) [reply src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs]: I think it could. I now return an empty list instead

File: src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs (hunk 9 above)
Comment (pbros) [reply src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs]: done. works better in GraphQL now too for casting :)

File: src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationInputObjectType.cs (hunk 10 above)
Comment (pbros) [reply src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationInputObjectType.cs]: done

File: src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs (hunk 11 above)
Comment (pbros) [reply src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs]: It returns an empty list if not results are found (the implementation calls ToListAsync() on an IEnumerable)

File: src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs [hunk 12]
```diff
@@ -13,6 +18,31 @@ public LocalizationQueryObjectType(IStringLocalizer<LocalizationQueryObjectType>
...
+                .Description(S["The localizations of the content item."])
+                .Argument<StringGraphType, string>("culture", "the culture of the content item")
+                .ResolveLockedAsync( async ctx =>
+                {
+                    var culture = ctx.GetArgument<string>("culture");
+                    var contentLocalizationManager = ctx.ResolveServiceProvider().GetService<IContentLocalizationManager>();
+
+                    if (culture != null)
+                    {
+                        var contentItems = new List<ContentItem>();
```
Comment (carlwoodhouse) [inline src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs]: you could save the list creation / adding here by using something like ...
```
var contentItem = await contentLocalizationManager.GetContentItemAsync(ctx.Source.LocalizationSet, culture);
return contentItem != null ? new[] { contentItem } : Enumerble.Empty<ContentItem>();
```

:)

Comment (pbros) [reply src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs]: nice and clean. made the changes :)

Comment (carlwoodhouse) [reply src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs]: nice :) also can remove `var contentItems = new List<ContentItem>();` now :) think you missed it

Comment (pbros) [reply src/OrchardCore.Modules/OrchardCore.ContentLocalization/GraphQL/LocalizationQueryObjectType.cs]: oops ;)

**Comment by carlwoodhouse** [comment]
LGTM
